### ISSUE TITLE
Mixed content issue for HTTP rel="next" links in Firefox

### ISSUE THREAD

**Issue by hwdsl2** [TryGhost/Ghost#6270]
### Issue Summary:

Firefox treats HTTP rel="next" links as "mixed content", blocks those and displays a yellow triangle on the green padlock in front of the URL.

### Steps to Reproduce:

1. In Firefox, visit https://hannah.wf/
2. Notice that there is a yellow triangle on the green padlock due to mixed content.
3. Press Shift+F5 to open developer tools. Click on "Console". The following is shown:
<pre><code>Blocked loading mixed active content "http://hannah.wf/page/2/"
</code></pre>
4. View page source and notice this line: `<link rel="next" href="http://hannah.wf/page/2/" />`

#### Expected result: 
Green padlock only.

#### Actual result: 
Green padlock with yellow triangle.

### Technical details:
Platforms where the issue observed:
* Ghost Version: 0.7
* Browser: Firefox Version 43

**Comment by howsonwhocodes** [comment]
@hwdsl2 - a work around is to set the blog url to https://www.example.com in config.js - if you use nginx and get stuck in a loop add proxy_set_header X-Forwarded-Proto $scheme; to your location configuration.

**Comment by hwdsl2** [comment]
@howsonwhocodes Thanks for the workaround.

**Comment by jtwebman** [comment]
@ErisDS The other way to go instead of forcing would be to always use relative paths for links to pages. It looks like most pages do just not the <head> tags and links as well as the <header> logo link and the rss feed link in the footer. I'll take a look and see if I can't fix this.

**Comment by jtwebman** [comment]
@ErisDS Actually it looks like there is already a way to fix this. In you config file you need to set:
```javascript
urlSSL: 'https://hannah.wf/',
```
This was the line that told me this: [core->cerver->config->url.js Line 21](https://github.com/TryGhost/Ghost/blob/master/core/server/config/url.js#L21)

**Comment by ErisDS** [comment]
`urlSSL` is used when you want to have a different URL for your SSL requests, e.g. secure.hannah.wf, it's pretty much only used for Heroku I think, which requires a different domain for secure requests.

If you set `url` to be http, then we treat this as the 'default'. Both http & https will work, but canonical urls are generated using http. In the case of the rel next/previous links, these are not canonical, and so we should probably generate them using the protocol of the request, rather than the default.

If you set `url` to be https, then we treat this as the default instead, but this also disables http entirely. 

In the case of my blog, I had temporarily disabled the forceSSL option on Ghost(Pro) which is what controls whether the protocol for `url` is https or http. I've enabled this again now, and the problem goes away.

The code for figuring out the base URL for any given location in Ghost, including which host and protocol to use, is spread across the codebase and needs unifying. We also need to be smarter about when a protocol needs to be the 'default' vs when it needs to match the request. I think there's still quite a few things to do here.

**Comment by jtwebman** [comment]
@ErisDS  From what I found everything goes through the config.urlFor call which eventually calls getBaseUrl function. I can make it work the way you described. So if urlSSL is not set then I will take off the http and add https if the user came in secure. 

As a side note Ghost does do a bunch of manual url manipulation what is the process of adding new dependencies? [urijs](https://www.npmjs.com/package/urijs) is a good library for doing this much easier.

**Comment by howsonwhocodes** [comment]
@jtwebman your method seems to work best for me and delivers the ability to server over http which is the best of both. Longer term there is a bug fix to be made but its a great work around

**Comment by ErisDS** [comment]
When it comes to adding dependencies, the process starts by making a case for adding one, either in the issue or as part of the PR write-up. 

The main two things to consider are, what benefit does it provide and why the particular dependency you suggest and not another one?
For quality, we look at whether there are alternatives, checking that the dep is MIT licensed and then looking for regular commits, high star rating and good test coverage.

In this case, I think urijs is *the* module for this, but it's worth doing a quick google to see if there are competitors rising up and worth taking note of.

When it comes to adding the dependency, we have a special commit format [example](https://github.com/TryGhost/Ghost/commit/c6ed823201557f14078e2641c4a24280e844fbfd), and it's preferred that a refactor to change code to use the new dependency is done separately to changes to functionality - so new dep & relevant changes = 1 PR, new functionality = second PR.

Hope that helps :) 

A side note regarding the behaviour change, if the protocol set in `url` is https, the protocol must always be https. If the protocol set in `url` is http, then the protocol may change to match the request if that is sensible. E.g. canonical urls should use the default. 

Bottom line is, the only time a mixed content warning should occur is by 'user error' E.g. if a user has hard-coded a http url into a post or image slot when the blog is set to or served over https. Ghost's internal mechanisms should always behave correctly.

**Comment by jtwebman** [comment]
I am close but I need to write a few tests as well as find how the navigation is getting urls as it doesn't seem to be switching to https. @ErisDS would it be ok to change navigation bar to use relative urls?

Here is my branch so far: https://github.com/jtwebman/Ghost/tree/bug_mixed_content_firefox

**Comment by ErisDS** [comment]
> would it be ok to change navigation bar to use relative urls?

Not sure what you mean here?

**Comment by jtwebman** [comment]
@ErisDS If a theme uses the navigation helper for handlebars to create a navigation bar like the default one created in the casper navigation bar:

```html
<body class="home-template nav-closed">

    <div class="nav">
    <h3 class="nav-title">Menu</h3>
    <a href="#" class="nav-close">
        <span class="hidden">Close</span>
    </a>
    <ul>
            <li class="nav-home nav-current" role="presentation"><a href="http://localhost:2368/">Home</a></li>
    </ul>
    <a class="subscribe-button icon-feed" href="http://localhost:2368/rss/">Subscribe</a>
</div>
``` 

The url values are absolute links. We should look to see if they are site urls and one ether switch them to https if they come in on ssl or remove the http://localhost:2368 and make the relative links like / and /rss/.

I am not sure this would be the same issue though. I could open another issue for that fix / discussion.

**Comment by ErisDS** [comment]
Those urls are all stored as relative paths in the database. They are output as absolute urls because that's what is defined in the [default template](https://github.com/TryGhost/Ghost/blob/6d42df029a2ca67b6fc390ea08625e5135fa100d/core/server/helpers/tpl/navigation.hbs). 

The `{{url}}` helper, with absolute set to true, calls `config.urlFor` [here](https://github.com/TryGhost/Ghost/blob/6f457768a2427ae70faa889f1debe3ad6cc455a5/core/server/helpers/url.js). 

There should be no need to change anything other than the behaviour of `urlFor` :)

**Comment by jtwebman** [comment]
@ErisDS Actually I think the way Ghost is doing the navigation helper that the url is switched from the helper to just a property just like the {{label}}. [here](https://github.com/TryGhost/Ghost/blob/master/core/server/helpers/navigation.js#L52) 

When I did a console.log on the url helper those navigation urls never go through it. Also my branch fixed the url helper as it is calling the config.urlFor wrong on this line. [here](https://github.com/TryGhost/Ghost/blob/master/core/server/helpers/url.js#L30) 

It should be passing three parameters like I changed it in my branch (Well 4 in mine because you have to have a way to tell the urlFor you want to use the default for absolute urls instead of replacing them). [here](https://github.com/jtwebman/Ghost/blob/bug_mixed_content_firefox/core/server/helpers/url.js#L31)

When I get home tonight I'll take a closer look and test fully what {{url}} is set to in the navigation.

**Comment by jtwebman** [comment]
@ErisDS You were correct in it was still calling the helper it was just passing the nav data and missing the secure. I updated the navigation helper to include the context.secure from the page. Thanks for the help! The PR above fixes all link issues besides if the theme uses {{@blog.url}} as that is directly tied to the config.url.

**Comment by jtwebman** [comment]
@ErisDS Is this issue still open? I believed it was fixed.

**Comment by ErisDS** [comment]
:+1:
